### ISSUE TITLE
Game controller support 

### ISSUE THREAD

**Issue by ctrl-meta-f** [diasurgical/devilutionX#67]
I believe playing Diablo using a game controller is a lot of fun!

SDL has decent support for game controllers. I managed to tweak the engine a bit to detect my Xbox One controller and handle some button presses, so the whole business looks quite promising.

However, it looks like introducing a good controller experience is a lot of work and I'd like to discuss the details in scope if this issue.

First, the event handling loop in the original code handles low-level IO events like mouse clicks and keypresses, not high-level actions like "move one step north-west" or "attack". So I guess one'll have to
  * (easy) add new message types like WM_MOVELEFT and WM_ATTACK and translate controller actions into these actions.
  * (tricky) write implementation for the new messages in diablo.cpp.

Second, I'm not really sure how to combine controller and mouse in a single UI. SDL uses Xbox 360 controller as a model, which has two sticks and a d-pad. I believe it makes sense to move the mouse using the right stick, and use the left stick to move the character around. Most of the actions though (opening doors/chests, attack, casting spells, opening map, etc.) should be possible to execute without ever moving the mouse. 

Third, finding a good mapping from buttons to game actions is an important problem by itself. One can use original Playstation mapping for inspiration
![original Playstation controller mapping](https://user-images.githubusercontent.com/46421162/54951373-bc2d4f80-4f43-11e9-9139-de2e725fcf90.png)
Diablo 3 runs on consoles so there is also a mapping, but I don't think it suits Diablo 1 well.
![xbox controls](https://user-images.githubusercontent.com/46421162/54951501-1b8b5f80-4f44-11e9-9443-bdd97e8e0297.jpg)

What do you think?

**Comment by AJenbo** [comment]
Hi @ctrl-meta-f 

We defiantly plan to have controller support. The PS and D3 sachems are a really good starting point,  thanks for points the mappings.
@erfg12 has already started on implementing a POC: https://github.com/erfg12/devilution
https://i.imgur.com/Tdt9MuP.mp4

At this point most the work is on making the UI easy to navigate with a dpad/keyboard like like device.

**Comment by ctrl-meta-f** [comment]
Thanks for a quick reply!

Looks like @erfg12 mostly solved the first (and the hardest) problem, though he uses XInput API instead of SDL (which suits the X engine more and should support more controllers).

Are there any plans to merge the Part 1 of his plan (moving with WASD, picking/attacking things nearby, etc.) into the devilutionX engine in the near future?

I could then use his work to polish my cross-platform SDL-based code and prepare a pull-request.

**Comment by erfg12** [comment]
sup. @ctrl-meta-f @AJenbo 

Yeah I'm putting the finishing touches on my Devilution keyboard extension and game controller project. This will use XInput since using the `SDL_Init(joystick || controller)` function makes the game freeze. Once this is done (soon hopefully) I will move onto DevilutionX and importing my code into there along with proper SDL controller functionality.

**Comment by erfg12** [comment]
Here's what I plan on using for the Xbox 360 controller:

GAMEPAD_A - SPACE
GAMEPAD_B - I key
GAMEPAD_X RETURN
GAMEPAD_Y - X key
GAMEPAD_BACK - TAB
GAMEPAD_START - ESCAPE
GAMEPAD_LEFT_SHOULDER - H key
GAMEPAD_RIGHT_SHOULDER - C key
GAMEPAD_DPAD - Arrow Keys

**Comment by ctrl-meta-f** [comment]
JFYI, here is the minimal diff to enable SDL game controller in DevolutionX:

```diff
diff --git a/SourceX/dx.cpp b/SourceX/dx.cpp
index 926d2b0..be2ba26 100644
--- a/SourceX/dx.cpp
+++ b/SourceX/dx.cpp
@ -19,6 +21,7 @@ char gbEmulate; // unread
 SDL_Window *window;
 SDL_Renderer *renderer;
 SDL_Texture *texture;
+SDL_GameController* game_controller;
 
 /** 32-bit in-memory backbuffer surface */
 SDL_Surface *surface;
@@ -443,6 +446,21 @@ void sdl_present_surface()
 	surface_dirty = false;
 }
 
+void sdl_init_controller() {
+    const int num_joysticks = SDL_NumJoysticks();
+
+    SDL_Log("Found %i joysticks", num_joysticks);
+
+    game_controller = 0;
+    for(int i = 0; i < num_joysticks; ++i) {
+        if (SDL_IsGameController(i)) {
+            SDL_Log("Using game controller #%i named \'%s\'", i,
+                    SDL_GameControllerNameForIndex(i));
+            game_controller = SDL_GameControllerOpen(i);
+        }
+    }
+}
+
 void j_lock_buf_priv(BYTE idx)
 {
 	j_unlock_buf_priv(idx); // what is idx?
diff --git a/SourceX/dx.h b/SourceX/dx.h
index da5db95..9f785d9 100644
--- a/SourceX/dx.h
+++ b/SourceX/dx.h
@@ -10,9 +10,12 @@ extern SDL_Texture *texture;
 extern SDL_Surface *surface;
 extern SDL_Palette *palette;
 extern SDL_Surface *pal_surface;
+extern SDL_GameController* game_controller;
 
 void sdl_present_surface();
 
 void sdl_update_entire_surface();
 
+void sdl_init_controller();
+
 }  // namespace dvl
diff --git a/SourceX/miniwin/misc.cpp b/SourceX/miniwin/misc.cpp
index 925388c..6640965 100644
--- a/SourceX/miniwin/misc.cpp
+++ b/SourceX/miniwin/misc.cpp
@@ -291,6 +291,8 @@ HWND CreateWindowExA(
 		flags |= SDL_WINDOW_INPUT_GRABBED;
 	}
 
+    sdl_init_controller();
+
 	window = SDL_CreateWindow(lpWindowName, SDL_WINDOWPOS_UNDEFINED, SDL_WINDOWPOS_UNDEFINED, nWidth, nHeight, flags);
 	atexit(FakeWMDestroy);
 
diff --git a/SourceX/miniwin/misc_msg.cpp b/SourceX/miniwin/misc_msg.cpp
index c03983a..484d681 100644
--- a/SourceX/miniwin/misc_msg.cpp
+++ b/SourceX/miniwin/misc_msg.cpp
@@ -3,6 +3,7 @@
 
 #include "devilution.h"
 #include "stubs.h"
+#include "dx.h"
 
 /** @file
  * *
@@ -13,6 +14,13 @@ namespace dvl {
 
 static std::deque<MSG> message_queue;
 
+static struct {
+    SDL_GameControllerButton button;
+    int dvl_key;
+} controller_keys_mapping[] = {
+    {SDL_CONTROLLER_BUTTON_DPAD_UP, DVL_VK_TAB}
+};
+
 static int translate_sdl_key(SDL_Keysym key)
 {
 	int sym = key.sym;
@@ -168,7 +191,40 @@ WINBOOL PeekMessageA(LPMSG lpMsg, HWND hWnd, UINT wMsgFilterMin, UINT wMsgFilter
 			return false_avail();
 		}
 	} break;
-	case SDL_MOUSEBUTTONUP: {
+    case SDL_CONTROLLERDEVICEADDED:
+    case SDL_CONTROLLERDEVICEREMOVED:
+    case SDL_CONTROLLERDEVICEREMAPPED:
+        sdl_init_controller();
+        break;
+    case SDL_CONTROLLERBUTTONDOWN:
+    case SDL_CONTROLLERBUTTONUP: {
+        int key = -1;
+        for (size_t i = 0; i < sizeof controller_keys_mapping / sizeof controller_keys_mapping[0]; ++i) {
+            if (SDL_GameControllerGetButton(game_controller, controller_keys_mapping[i].button)) {
+                key = controller_keys_mapping[i].dvl_key;
+            }
+        }
+        if (key >= 0) {
+            lpMsg->message = e.type == SDL_CONTROLLERBUTTONDOWN ? DVL_WM_KEYDOWN : DVL_WM_KEYUP;
+            lpMsg->wParam = (DWORD)key;
+        } else {
+            DUMMY_PRINT("Unkwnown controller event");
+        }
+    } break;
+    case SDL_MOUSEBUTTONUP: {
 		int button = e.button.button;
 		if (button == SDL_BUTTON_LEFT) {
 			lpMsg->message = DVL_WM_LBUTTONUP;
```

> GAMEPAD_DPAD - Arrow Keys

I believe using the left stick (GAMEPAD_LEFT_THUMB) is much more convenient, especially for diagonal moves.

**Comment by neuromancer** [comment]
This patch is failing in the latest revision (cb9ef7a8b8c30e9a3042be39cb13fd041fa3b04)

**Comment by AJenbo** [comment]
@neuromancer does it fail to apply or give an error?

**Comment by neuromancer** [comment]
This is the result after renaming `SourceX/dx.h` by `Source/dx.h` in the patch (since the file was moved)
 
```
$ patch -p1 < joy.diff 
patching file SourceX/dx.cpp
Hunk #1 FAILED at 443 (different line endings).
1 out of 1 hunk FAILED -- saving rejects to file SourceX/dx.cpp.rej
patching file Source/dx.h
Hunk #1 FAILED at 10 (different line endings).
1 out of 1 hunk FAILED -- saving rejects to file Source/dx.h.rej
patching file SourceX/miniwin/misc.cpp
Hunk #1 FAILED at 291 (different line endings).
1 out of 1 hunk FAILED -- saving rejects to file SourceX/miniwin/misc.cpp.rej
patching file SourceX/miniwin/misc_msg.cpp
patch: **** unexpected end of file in patch
```

**Comment by AJenbo** [comment]
I updated the patch so that it stays out of the direct draw code, use `git apply joy.diff` to apply it.

**Comment by neuromancer** [comment]
For some reason, it fails: 

```
$ git apply joy.diff 
error: patch failed: SourceX/miniwin/misc.cpp:30
error: SourceX/miniwin/misc.cpp: patch does not apply
```

Are you sure you used cb9ef7a8b8c30e9a3042be39cb13fd041fa3b045 ?

**Comment by AJenbo** [comment]
Yes, and it still works for me, but just in case the issue is line endings or some other copy paste issue here it is as an attachment:
[joy.zip](https://github.com/diasurgical/devilutionX/files/3360697/joy.zip)

**Comment by AJenbo** [comment]
This is now mostly implemented in https://github.com/diasurgical/devilutionX/pull/181 but some further work is needed before it will be merged into master. Note that this branch is currently broken on Mac.

**Comment by neuromancer** [comment]
> Can confirm that this actually works, not that it's of much use without additional mapping and some way to control the pointer.

Is that solved in #181 ?

**Comment by AJenbo** [comment]
Yes, though there might still be some aspects that need some work.

**Comment by AJenbo** [comment]
If any one is able to extract the Game Controller code from https://github.com/diasurgical/devilutionX/pull/181 and clean it up so that is is ready for merge that would be a big help.

**Comment by neuromancer** [comment]
#181 needs a lot more of work to be merged? I tested a few months ago and it was working fine (in Linux).

**Comment by AJenbo** [comment]
@neuromancer I was actually just working on rebasing it on the latest master branch. Some of the issues where solved with in the past couple of weeks and the diff dropped by about 50 lines. But as you can see from it's check list there are still some outstanding issues that needs to be resolved before it can be merged.

The big issue is that it isn't clear where the Nintendo Swith platform code ends and where the game controller code start. I know that there are several features of the game controller code that are simply disabled and only the basics are functional on other platforms. There are likely also parts of the regular functionality that has been disabled or isn't performing correctly as a result of some of the changes that where done (you probably won't notice on non-Switch systems). All of this makes it over all very hard to maintain and as such it's not ready to be merged in to the main branch. The whole process isn't helped by me not having a Switch either, I do however have a game controller but probably won't have time to work a lot on it before we land 1.0.0 and Hellfire.

In case you have time it would be helpful if you can test that things are still working the same in my rebase branch: https://github.com/diasurgical/devilutionX/pull/271

**Comment by glebm** [comment]
I've made some Diablo PS1 screenshots with commentary: https://imgur.com/gallery/aXe0Vor

The biggest thing the PS1 version has in terms of controller support is Auto-Aim

**Comment by neuromancer** [comment]
@AJenbo you still need testing for #329 ? (#271 was closed/deleted)

**Comment by AJenbo** [comment]
@neuromancer a bit more testing on #329 would be nice, #271 has already become the new #181

**Comment by neuromancer** [comment]
I just tested the first level on #329. Taking with certain NPCs and melee combat works but IMHO the controller support needs a button to interact with objects (or certain people) nearby the player (when they are selected/identified). It is possible to click on objects manually using the controller, but it is too slow.  Perhaps the Auto-Aim that @glebm mentions is a good solution.

**Comment by AJenbo** [comment]
Nice, thank you. We are going to tackle controller support/features as a separate issue. That way we can merge the Nintendo Switch code and then work on controller support for all platforms so that everyone benefits from improvements like the one you just mentioned.

**Comment by glebm** [comment]
Controller support is fully implemented in #395, give it a try.

Controls:

- Left analog / DPad: move hero
- Right analog: simulate mouse
- A: attack nearby enemies, talk to towns people and merchants, pickup & drop items in inventory, OK while in main menu
- B: Select spell, cancel while in main menu
- X: pickup gold, potions & equipment from ground, open chests and doors that are nearby, use item when in inventory (useful to read books etc.)
- Y: cast spell, go to previous screen when talking to people and in shops, delete character while in main menu
- R1: inventory
- L1: character
- R2: drink mana potion
- L2: drink health potion
- Left analog click: quest log
- Right analog click: left mouse click
- Select: automap
- Start: game Menu, skip intro

**Comment by neuromancer** [comment]
You did a awesome job @glebm! I tested your branch for a few minutes, in the first dungeon and looks like it is definitely working as expected during combat (which was a big issue in the previous version of this feature). My only concern so far is the default for the Steam Controller (which I'm using, emulated as an xbox controller using sc-controller). When I started the game and arrived in the main menu, pressing the A button will close the game without any warning.

**Comment by AJenbo** [comment]
I also had an unexpected shutdown on the RG300, not sure if that is related, don't really have any debug info though.

Another thing though is that it is a bit odd that the attack button also interacts with doors since that can make combat hard when you are close to a door, and seams unnecessary since you can use the "use" button also.

**Comment by glebm** [comment]
Another thing to improve is the top button should use the focused iventory item. Currently there is no way to read spell books.

**Comment by neuromancer** [comment]
> Another thing though is that it is a bit odd that the attack button also interacts with doors since that can make combat hard when you are close to a door, and seams unnecessary since you can use the "use" button also.

I can confirm this is an issue when fighting with a large of enemies in small rooms . 

Additionally, sometimes the attack button will make you face enemies outside the screen, the hit "near" enemies is not always working as expected (perhaps the scope of this feature should be possible to configure, depending on how confident you are?)

**Comment by AJenbo** [comment]
The check for distance is fairly primitive, it will probably target visible yet unreachable enemies (behind gate), i have made some suggestions for how to improve it.

**Comment by AJenbo** [comment]
Basic support is now merged to the main branch, thanks to @glebm @rsn8887 and @erfg12 for there part in this work.

**Comment by AJenbo** [comment]
Further improvement have been moved to this issue https://github.com/diasurgical/devilutionX/issues/418
